mllib/src/main/scala/org/apache/spark/mllib/evaluation/MultilabelMetrics.scala

```diff
@@ -38,85 +40,65 @@ class MultilabelMetrics @Since("1.2.0") (predictionAndLabels: RDD[(Array[Double]
       (r.getSeq[Double](0).toArray, r.getSeq[Double](1).toArray)
     })
 
-  private lazy val numDocs: Long = predictionAndLabels.count()
+  /**
+   * Use MultilabelSummarizer to calculate all summary statistics of predictions
+   * and labels on one pass.
+   */
+  private val summary: MultilabelSummarizer = {
+    predictionAndLabels
+      .treeAggregate(new MultilabelSummarizer)(
+        (summary, sample) => summary.add(sample._1, sample._2),
+        (sum1, sum2) => sum1.merge(sum2)
+      )
+  }
 
-  private lazy val numLabels: Long = predictionAndLabels.flatMap { case (_, labels) =>
-    labels}.distinct().count()
 
   /**
    * Returns subset accuracy
    * (for equal sets of labels)
    */
   @Since("1.2.0")
-  lazy val subsetAccuracy: Double = predictionAndLabels.filter { case (predictions, labels) =>
-    predictions.deep == labels.deep
-  }.count().toDouble / numDocs
+  lazy val subsetAccuracy: Double = summary.subsetAccuracy
```

Comment:
These shouldn't be lazy now. They just delegate to a non-lazy object that's already initialized.